x/zkauth/ante/zkauth.go

```diff
@@ -27,25 +31,123 @@ func (zka ZKAuthMsgDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bo
 		Basically, in the case of zkauth msg, ephPubKey must be idempotent for each msg.
 	*/
 
+	isZKAuthTx, zkMsgs, pubKeys, err := isZKAuthTx(tx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !isZKAuthTx {
+		svd := authante.NewSigVerificationDecorator(zka.ak, zka.signModeHandler)
+		return svd.AnteHandle(ctx, tx, simulate, next)
+	}
+
+	for i, zkMsg := range zkMsgs {
+		// verify ZKAuth signature
+		if err := zkauthtypes.VerifyZKAuthSignature(ctx, zka.zk, pubKeys[i].Bytes(), zkMsg); err != nil {
+			return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "invalid zkauth signature, %s", err)
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+type ZKAuthSetPubKeyDecorator struct {
+	zk zkauthtypes.ZKAuthKeeper
+	ak authante.AccountKeeper
+}
+
+func NewZKAuthSetPubKeyDecorator(zk zkauthtypes.ZKAuthKeeper, ak authante.AccountKeeper) ZKAuthSetPubKeyDecorator {
+	return ZKAuthSetPubKeyDecorator{zk: zk, ak: ak}
+}
+
+func (zsp ZKAuthSetPubKeyDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	isZKAuthTx, zkMsgs, _, err := isZKAuthTx(tx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !isZKAuthTx {
+		spk := authante.NewSetPubKeyDecorator(zsp.ak)
+		return spk.AnteHandle(ctx, tx, simulate, next)
+	}
+
+	for _, zkMsg := range zkMsgs {
+		msgs, err := zkMsg.GetMessages()
+		if err != nil {
+			return ctx, err
+		}
+		for _, msg := range msgs {
+			for _, signer := range msg.GetSigners() {
+				zsp.ak.SetAccount(ctx, zsp.ak.NewAccountWithAddress(ctx, signer))
+			}
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+type ZKAuthIncrementSequenceDecorator struct {
+	ak authante.AccountKeeper
+}
+
+func NewIncrementSequenceDecorator(ak authante.AccountKeeper) ZKAuthIncrementSequenceDecorator {
+	return ZKAuthIncrementSequenceDecorator{
+		ak: ak,
+	}
+}
+
+func (zkisd ZKAuthIncrementSequenceDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	isZKAuthTx, zkMsgs, _, err := isZKAuthTx(tx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !isZKAuthTx {
+		isd := authante.NewIncrementSequenceDecorator(zkisd.ak)
+		return isd.AnteHandle(ctx, tx, simulate, next)
+	}
+
+	for _, zkMsg := range zkMsgs {
+		msgs, err := zkMsg.GetMessages()
+		if err != nil {
+			return ctx, err
+		}
+		for _, msg := range msgs {
+			for _, signer := range msg.GetSigners() {
+				acc := zkisd.ak.GetAccount(ctx, signer)
+				if err := acc.SetSequence(acc.GetSequence() + 1); err != nil {
+					panic(err)
+				}
+
+				zkisd.ak.SetAccount(ctx, acc)
+			}
+		}
```

Comment:
Since the signer of zkMsgs can be obtained directly from zkMsg, how about use it?
```suggestion
		for _, signer := range zkMsg.GetSigners() {
			acc := zkisd.ak.GetAccount(ctx, signer)
			if err := acc.SetSequence(acc.GetSequence() + 1); err != nil {
				panic(err)
			}

			zkisd.ak.SetAccount(ctx, acc)
		}
```

Reply:
Are the signers of zkMsg EphPubKey? The address I want to set now is for the zkauth address, so I'm trying to get it from the msg signer.

Reply:
No, they aren't. We can get the zkauth address by zkMsg.GetSigners.
https://github.com/Finschia/finschia-sdk/blob/ca6ca76ea1eb5d3aff74bb8fe8910445c3b456ed/x/zkauth/types/msgs.go#L96-L103

Reply:
I understood, I fix it.